Shared multiheap: extensions. Add user_data and sys_multi_heap_get_heap

marcinszkudlinski opened the pull request.

dbaluta commented on include/sys/multi_heap.h:
```diff
@@ -51,10 +51,16 @@ struct sys_multi_heap;
 typedef void *(*sys_multi_heap_fn_t)(struct sys_multi_heap *mheap, void *cfg,
 				     size_t align, size_t size);
 
```
Looks like you need to fix the formatting for you commit message. Please add a new line between subject and body. Also, please try to trim the commit message line length at around 100 chars.

andyross commented on include/sys/multi_heap.h:
```diff
@@ -51,10 +51,16 @@ struct sys_multi_heap;
+struct sys_heap_ext {
+	struct sys_heap *sysheap;
+	void *metadata;
+};
```
Calling this "_ext" is a little confusing.  Usually that implies that it's an "extension" to the previous record, but this is a separate data structure that associates a heap with a user-supplied pointer.

Also, it's a 100% multiheap data structure, it's not part of sys_heap.  So I'd suggest naming more like:

```c
struct sys_multi_heap_rec {
    struct sys_heap *heap;
    void *user_data;
};
```

andyross commented on include/sys/multi_heap.h:
```diff
@@ -125,6 +132,17 @@ void *sys_multi_heap_alloc(struct sys_multi_heap *mheap, void *cfg, size_t bytes
+ * @param mheap Multi heap pointer
+ * @param addr address to be found
+ */
+struct sys_heap_ext *sys_multi_heap_get_heap(struct sys_multi_heap *mheap, void *addr);
```
Returning pointers into internally-managed data structures is sort of a dangerous pattern.  Suggest making this const.

Alternatively, since it's just a two-word struct, you could pass it as an argument and have the multiheap code fill it in for the caller.  That would open the possibility of returning an error code like -ENOENT on bad data.

andyross commented on include/sys/multi_heap.h:
```diff
@@ -125,6 +132,17 @@ void *sys_multi_heap_alloc(struct sys_multi_heap *mheap, void *cfg, size_t bytes
+ * @brief Get a specific heap for provided address
+ *
+ * Returns the pointer to a single system heap (with metadata)
+ *  controlling the provided pointer
```
Pedantry: we should probably specify that the pointer must be an un-freed return value from sys_multi_heap_alloc().  It's true that right now, it's just a search of address ranges so it will work for any address inside the heap memory.  But it's easy to imagine a more complicated scheme showing up in the future where there's a header in the block or something.

andyross reviewed (commented):
Some nitpicks on API, but on the whole this looks very reasonable to me.

ceolin commented on lib/os/multi_heap.c:
```diff
@@ -49,9 +51,10 @@ void *sys_multi_heap_aligned_alloc(struct sys_multi_heap *mheap,
 }
 
-void sys_multi_heap_free(struct sys_multi_heap *mheap, void *block)
+int sys_multi_heap_get_heap(struct sys_multi_heap *mheap, void *addr,
```
`const struct sys_multi_heap *mheap` ?

marcinszkudlinski replied:
const struct is  good as well, but in case the heap has not been found NULL will be returned. The higher layer need to check it - and it's easier to overlook it than checking if the exit code is != 0 
can easily change it, its up to you and @andyross ?

ceolin commented:
looks good to me the changes.

andyross reviewed (dismissed):
Refresh review, looks good to me

nashif commented:
@marcinszkudlinski  please fix the checkpatch issues:

```
-:47: WARNING:SPACE_BEFORE_TAB: please, no space before tabs
#47: FILE: include/sys/multi_heap.h:144:
+ * ^I^I^I or NULL if the heap has not been found$

-:49: WARNING:LONG_LINE: line length of 105 exceeds 100 columns
#49: FILE: include/sys/multi_heap.h:146:
+const struct sys_multi_heap_rec *sys_multi_heap_get_heap(const struct sys_multi_heap *mheap, void *addr);

-:99: WARNING:LONG_LINE: line length of 104 exceeds 100 columns
#99: FILE: lib/os/multi_heap.c:54:
+const struct sys_multi_heap_rec *sys_multi_heap_get_heap(const struct sys_multi_heap *mheap, void *addr)

-:121: WARNING:BLOCK_COMMENT_STYLE: Block comments use a trailing */ on a separate line
#121: FILE: lib/os/multi_heap.c:76:
+	 * FIXME: return -ENOENT when a proper heap is not found */

- total: 0 errors, 4 warnings, 1[35](https://github.com/zephyrproject-rtos/zephyr/runs/5481716450?check_suite_focus=true#step:9:35) lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.

Your patch has style problems, please review.
```

marcinszkudlinski committed:
lib/os: Add metadata to heap in multi_heap

When operating on different kinds of heaps sometimes there's a need to
perform special operations on heap, poweroff memory bank when releasing
memory etc. Therefore some additional data may be required.
Metadata is a point to keep such data.

Signed-off-by: Marcin Szkudlinski <marcin.szkudlinski@intel.com>

marcinszkudlinski committed:
lib/os: Add sys_multi_heap_get_heap function for multi_heap

Sometimes there's a need to perform special operations on a heap
for a specified block. I.e. release virtual memory mapping,
power off memory bank etc.
The added procedure gets a pointer to a proper heap with metadata
for a given address.

Signed-off-by: Marcin Szkudlinski <marcin.szkudlinski@intel.com>
